### PR TITLE
P1659R3 starts_with and ends_with

### LINKED ISSUE #4647: [2021-06 LWG Motion 10] P1659R3 starts_with and ends_with
[P1659R3](https://wg21.link/p1659r3)

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #4647
Fixes cplusplus/papers#442

### CODE REVIEW

File: source/algorithms.tex [hunk 1]
```diff
@@ -1131,6 +1131,36 @@
     constexpr ForwardIterator
       search(ForwardIterator first, ForwardIterator last, const Searcher& searcher);
 
+  namespace ranges {
+    // \ref{alg.starts.with}, starts_with
+    template<@\libconcept{input_iterator}@ I1, @\libconcept{sentinel_for}@<I1> S1, @\libconcept{input_iterator}@ I2, @\libconcept{sentinel_for}@<I2> S2,
+             class Pred = ranges::equal_to, class Proj1 = identity, class Proj2 = identity>
+      requires indirectly_comparable<I1, I2, Pred, Proj1, Proj2>
```
Comment (JohelEGP) [inline source/algorithms.tex:1138]:
```suggestion
      requires @\libconcept{indirectly_comparable}@<I1, I2, Pred, Proj1, Proj2>
```

File: source/algorithms.tex [hunk 2]
```diff
@@ -1131,6 +1131,36 @@
...
+      constexpr bool starts_with(I1 first1, S1 last1, I2 first2, S2 last2, Pred pred = {},
+                                 Proj1 proj1 = {}, Proj2 proj2 = {});
+    template<@\libconcept{input_range}@ R1, @\libconcept{input_range}@ R2, class Pred = ranges::equal_to,
+             class Proj1 = identity, class Proj2 = identity>
+      requires @\libconcept{indirectly_comparable}@<iterator_t<R1>, iterator_t<R2>, Pred, Proj1, Proj2>
+      constexpr bool starts_with(R1&& r1, R2&& r2, Pred pred = {},
+                                 Proj1 proj1 = {}, Proj2 proj2 = {});
+
+    // \ref{alg.ends.with}, ends_with
```
Comment (JohelEGP) [inline source/algorithms.tex:1147]:
```suggestion
    // \ref{alg.ends.with}, ends with
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

**Comment by jensmaurer** [comment]
@JohelEGP , thanks, fixed.

File: source/algorithms.tex [hunk 3]
```diff
@@ -4219,6 +4249,81 @@
...
+template<@\libconcept{input_iterator}@ I1, @\libconcept{sentinel_for}@<I1> S1, @\libconcept{input_iterator}@ I2, @\libconcept{sentinel_for}@<I2> S2,
+         class Pred = ranges::equal_to, class Proj1 = identity, class Proj2 = identity>
+  requires @\libconcept{indirectly_comparable}@<I1, I2, Pred, Proj1, Proj2>
+  constexpr bool ranges::starts_with(I1 first1, S1 last1, I2 first2, S2 last2, Pred pred = {},
+                                     Proj1 proj1 = {}, Proj2 proj2 = {});
+template<@\libconcept{input_range}@ R1, @\libconcept{input_range}@ R2, class Pred = ranges::equal_to, class Proj1 = identity,
+         class Proj2 = identity>
+  requires @\libconcept{indirectly_comparable}@<iterator_t<R1>, iterator_t<R2>, Pred, Proj1, Proj2>
+  constexpr bool ranges::starts_with(R1&& r1, R2&& r2, Pred pred = {},
+                                     Proj1 proj1 = {}, Proj2 proj2 = {});
```
Comment (tkoeppe) [inline source/algorithms.tex:4265]: @jwakely: I'm not sure why this second overload doesn't have a specification, but that's what's in the paper. Please let me know if anything is missing here (or if this is covered by some blanket wording elsewhere).

Comment (jensmaurer) [reply source/algorithms.tex:4265]: The range overloads map to the first/last overloads per [algorithms.requirements] p14.

**Pull request merged**